src/openrct2/peep/peep.c

```diff
@@ -2930,10 +2930,29 @@ static void peep_update_ride_sub_state_2_rejoin_queue(rct_peep* peep, rct_ride*
 		return;
 	}
 
+	uint16 slow = current_last;
+	bool increment = true;
 	rct_peep* queue_peep;
 	for (queue_peep = GET_PEEP(current_last);
 		queue_peep->next_in_queue != SPRITE_INDEX_NULL;
-		queue_peep = GET_PEEP(queue_peep->next_in_queue));
+		queue_peep = GET_PEEP(queue_peep->next_in_queue))
+	{
+		// Sometimes queuing peeps will form a loop, effectively making this `for`
+		// spin infinitely. Following code will break out of such loop.
+		increment = !increment;
+		if (increment) {
+			slow = queue_peep->next_in_queue;
+		}
+		if (queue_peep->next_in_queue == slow) {
```

Comment:
Yes, it looks wrong indeed. The check perhaps should happen before the assignment.